### PR TITLE
Mail stats

### PR DESCRIPTION
**Pull request by divyap1**
Calculating the amount of mail per route for both total figures, and handling moving back and forward for browsing past events has been implemented

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -15,13 +15,13 @@
...
 
 // Disable scroll for whole site
-html {
-  width: 100%;
-  height: 100%;
-  clip: auto;
-  position: absolute;
-  overflow: hidden;
-}
+// html {
```
Comment (TanaTanoi) [inline app/assets/stylesheets/application.scss]: remove dis comment biz before merge into master, or leave a comment if it needs to be dealt with elsewhere

File: app/controllers/business_stats_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class BusinessStatsController < ApplicationController
+  def amount_of_mail
+    if params[:id] == '0'
+      @event = Event.last
+    else
+      @event = Event.find(params[:id])
+    end
+
+    @all_mail = Event.where('id <= ?', @event.id).where('type == ?', 'CreateMail')
```
Comment (TanaTanoi) [inline app/controllers/business_stats_controller.rb]: I _think_ you can do `.where(type: 'CreateMail')`, which equates to the same thing (but looks nice lol)

File: app/controllers/business_stats_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    @data = {}
+    @all_mail.each do |mail|
```
Comment (TanaTanoi) [inline app/controllers/business_stats_controller.rb]: There's a super cool enumerator called `inject` which we can use here.
Basically, *inject* has two parameters, an *accumulator* (the return value that is kept during the loop) and a *value* (the same as `each`). It also takes an *initial* value for the accumulator as a parameter. This means we can do:
`@data = @all_mail.inject({}) do |data, mail|`
`[the inside of the code]`
`data`
`end`
So the parameter is an empty hash, which gets assigned to the `data` parameter, then `mail` is the same as in the `each` . The reason we put `data` at the end of the loop is because the accumulator (`data` in this case) becomes whatever the loop returns.

Overall, you don't _have_ to do this (it saves one line lol), but inject is really awesome and worth learning, hit me up if you have any questions about it :smile:

File: app/controllers/company_routes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    @discontinued_routes = CompanyRoute.where(discontinued: true)
+  end
+
+  def new
+    @company_route = CompanyRoute.new
+    @locations = Location.all  #TODO make unique return
+  end
+
+  def create
+    if params[:company_route][:origin_id].blank?
```
Comment (TanaTanoi) [inline app/controllers/company_routes_controller.rb]: maybe make this into a service that returns a hash like `{success: false, message: 'Destination location must be specified'}`, then you only have one if statement that dictates if you need to `redirect_back` or not.

File: app/controllers/company_routes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        redirect_back_with_alert('This route already exists')
+      elsif @company_route.save!
+        @company_route.update(discontinued: false)
+        @company_route.calculate_delivery_times
+        CreateUpdateCostEvent.new(@company_route).call
+
+        @kps_route = KpsRoute.where(
+                          origin: @company_route.origin,
+                          destination: @company_route.destination,
+                          priority: @company_route.kps_route_priority).first
```
Comment (TanaTanoi) [inline app/controllers/company_routes_controller.rb]: I _think_ the convention for multi line bracket blocks is to treat the brackets like braces, such that the open and close bracket are on the same indentation level.

tl;dr new line before the `).first` :stuck_out_tongue_winking_eye:

File: app/controllers/company_routes_controller.rb [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    @company_route = CompanyRoute.find(params[:id])
+  end
+
+  def reinstate
+    @company_route = CompanyRoute.find(params[:company_id])
+    if @company_route.update(discontinued: false)
+      CreateUpdateCostEvent.new(@company_route).call
+      redirect_to company_routes_path, notice: 'Route successfully reinstated'
+    else
+      redirect_back_with_alert('An error occured. Please try again.')
```
Comment (TanaTanoi) [inline app/controllers/company_routes_controller.rb]: I appreciate that you're consistently using single quotes for strings, I fail to do this :worried:

File: app/controllers/kps_routes_controller.rb [hunk 7]
```diff
@@ -1,46 +1,49 @@
...
 
   def create
     @kps_route = KpsRoute.new(kps_route_params)
-    if @kps_route.save!
-      # new KPS route event
-      redirect_to @kps_route
+    @same_route = KpsRoute.where(
+                      origin: @kps_route.origin,
+                      destination: @kps_route.destination,
+                      priority: @kps_route.priority)
```
Comment (TanaTanoi) [inline app/controllers/kps_routes_controller.rb]: same thang as before

File: app/models/company_route.rb [hunk 8]
```diff
@@ -15,4 +18,44 @@ class CompanyRoute < ActiveRecord::Base
   validates :frequency, presence: true
   validates :duration, presence: true
   validates :day, presence: true, inclusion: {in: Constants::DAYS}
+
+  def set_default_discontinued
+    self.discontinued || false
```
Comment (TanaTanoi) [inline app/models/company_route.rb]: nice

File: app/models/company_route.rb [hunk 9]
```diff
@@ -15,4 +18,44 @@ class CompanyRoute < ActiveRecord::Base
...
+  end
+
+  def kps_route_priority
+    priority = 'International '
+    if origin.country == destination.country
+      priority = 'Domestic '
+    end
+    if priority_type == 'Air'
```
Comment (TanaTanoi) [inline app/models/company_route.rb]: make these constants (Air, Domestic, International, etc.), I know it sounds silly cos it'll probably be more characters, but its worth it :stuck_out_tongue: plus they'll probably get used elsewhere

File: app/models/company_route.rb [hunk 10]
```diff
@@ -15,4 +18,44 @@ class CompanyRoute < ActiveRecord::Base
...
+    start_time = DateTime.parse(start_day.to_s) + 18.hours
+    current_time = start_time
+
+    while current_time < start_time + 7.days
+      new_day = current_time.strftime('%A')
+      new_time = Time.parse(current_time.to_s)
+      @delivery = DeliveryTime.create!(
+                                  company_route: self,
+                                  day: new_day,
+                                  time: new_time)
```
Comment (TanaTanoi) [inline app/models/company_route.rb]: bracket new line thingy

File: app/models/company_route.rb [hunk 11]
```diff
@@ -15,4 +18,44 @@ class CompanyRoute < ActiveRecord::Base
...
+      current_time = current_time + freq.hours
+    end
+  end
+
+  def unique_name
+    company + ': ' + origin.location_name + ' to ' + destination.location_name + ' by ' + priority_type
```
Comment (TanaTanoi) [inline app/models/company_route.rb]: nice :+1:

File: app/models/kps_route.rb [hunk 12]
```diff
@@ -10,4 +10,7 @@ class KpsRoute < ActiveRecord::Base
   validates :volume_price, presence: true
   validates :priority, presence: true, inclusion: {in: Constants::PRIORITY}
 
+  def unique_name
+    origin.location_name + ' to ' + destination.location_name + ' by ' + priority
```
Comment (TanaTanoi) [inline app/models/kps_route.rb]: this one too

File: app/models/location.rb [hunk 13]
```diff
@@ -1,4 +1,8 @@
 class Location < ActiveRecord::Base
   validates :city, presence: true
   validates :country, presence: true
+
+  def location_name
+    "#{city}, #{country}"
```
Comment (TanaTanoi) [inline app/models/location.rb]: nice :+1:

File: app/views/company_routes/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+          <th>Priority</th>
+          <th></th>
+        </thead>
+        <tbody>
+          <% @kps_routes.each do |route| %>
+          <tr>
+            <td><%= route.origin.location_name %></td>
+            <td><%= route.destination.location_name %></td>
+            <td><%= route.priority %></td>
+            <td><%= "$#{number_with_precision(route.weight_price, precision: 2)}" %></td>
```
Comment (TanaTanoi) [inline app/views/company_routes/index.html.erb]: maybe make a helper method for this in `company_routes_helper.rb` where you just enter the `weight_price` and it formats it, including the `$`

File: config/initializers/assets.rb [hunk 15]
```diff
@@ -3,9 +3,11 @@
 # Version of your assets, change this if you want to expire all your assets.
 Rails.application.config.assets.version = '1.0'
 
+
```
Comment (TanaTanoi) [inline config/initializers/assets.rb]: prob delete the new lines in this file so it doesn't show up in the diff or commit logs, but not a big issue

File: app/assets/stylesheets/business_stats.scss [hunk 16]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the BusinessStats controller here.
```
Comment (Recursively) [inline app/assets/stylesheets/business_stats.scss]: empty scss files are best removed

File: app/controllers/business_stats_controller.rb [hunk 17]
```diff
@@ -0,0 +1,25 @@
+class BusinessStatsController < ApplicationController
+  def amount_of_mail
+    if params[:id] == '0'
```
Comment (Recursively) [inline app/controllers/business_stats_controller.rb]: How come the id here is 0? I am assuming this is the case when an id isn't provided?

Unless this 0 is being passed explicitly for some reason, I think it would make more sense to have no id param passed and then do something like

`@event = params[:id].present? ? Event.find(params[:id]) : Event.last`

Comment (Recursively) [inline app/controllers/business_stats_controller.rb]: Although maybe not a ternary as I don't like the double `?`

File: app/controllers/business_stats_controller.rb (hunk 2 above)
Comment (Recursively) [inline app/controllers/business_stats_controller.rb]: This might be worth turning into a scope

File: app/controllers/business_stats_controller.rb (hunk 3 above)
Comment (Recursively) [inline app/controllers/business_stats_controller.rb]: +1

File: app/controllers/company_routes_controller.rb [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+          redirect_to company_routes_path, notice: "Route added does not have a price."
+        end
+      else
+        redirect_back_with_alert('An error occured. Please try again.')
+      end
+    end
+  end
+
+  def destroy
+    @company_route = CompanyRoute.find(params[:id])
```
Comment (Recursively) [inline app/controllers/company_routes_controller.rb]: `destroy, update, edit, show and reinstate` all have the same `@company_route = CompanyRoute.find(params[:id])` at the start of their calls.

Adding a `before_action :set_company_route, only: [:destroy, :update... etc]` would dry this out a bit.

`def set_company_route
  @company_route = CompanyRoute.find(params[:id])
end`

File: app/controllers/events_controller.rb [hunk 19]
```diff
@@ -4,6 +4,10 @@ def index
   end
 
   def show
-    @event = Event.find_by_id(params[:id])
+    if params[:id] == '0'
```
Comment (Recursively) [inline app/controllers/events_controller.rb]: Same as above, is 0 being passed explicitly for some reason?

File: app/controllers/kps_routes_controller.rb [hunk 20]
```diff
@@ -1,46 +1,49 @@
...
-    @kps_routes = KpsRoute.all
-  end
-
   def new
     @kps_route = KpsRoute.new
+    @company_route = CompanyRoute.find(params[:company_id])
   end
 
   def create
     @kps_route = KpsRoute.new(kps_route_params)
```
Comment (Recursively) [inline app/controllers/kps_routes_controller.rb]: Again a `before_action` would help dry this out too

File: app/helpers/business_stats_helper.rb [hunk 21]
```diff
@@ -0,0 +1,2 @@
+module BusinessStatsHelper
```
Comment (Recursively) [inline app/helpers/business_stats_helper.rb]: Can delete empty helpers

File: app/models/company_route.rb [hunk 22]
```diff
@@ -1,8 +1,11 @@
 class CompanyRoute < ActiveRecord::Base
+  after_initialize :set_default_discontinued
+
   has_many :delivery_times, foreign_key: :company_route_id
 
   belongs_to :origin, class_name: 'Location'
   belongs_to :destination, class_name: 'Location'
+  belongs_to :kps_route, class_name: 'KpsRoute'
 
   validates :company, presence: true
```
Comment (Recursively) [inline app/models/company_route.rb]: Validations for the same thing can be truncated, ie:

```ruby
validates :company, :origin, :frequency, :duration, presence: true
```

File: app/views/business_stats/amount_of_mail.html.erb [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+    <th>Origin</th>
+    <th>Destination</th>
+    <th>Total Volume</th>
+    <th>Total Weight</th>
+    <th>Total Number of Items</th>
+  </thead>
+  <tbody>
+    <% @data.each do |key, value| %>
+    <tr>
+      <td><%= key[0].location_name %></td>
```
Comment (Recursively) [inline app/views/business_stats/amount_of_mail.html.erb]: Both the key and value is an array here?

If that's the case, I feel like the key should be wrapped in something, even if it's only like a struct or something.

Also Rails has handy helper methods `.first` `.second` `.third` which can be used in place of the array access notation.

File: spec/controllers/business_stats_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
```
Comment (Recursively) [inline spec/controllers/business_stats_controller_spec.rb]: All these empty spec files, what are we doing with them @TanaTanoi ? Should we just remove them all, and recreate them if we do actually write specs?

**Comment by Recursively** [comment]
Had a look through this, and at @TanaTanoi's code review. Added a few extra comments as well, but really good work @divyap1 💯

File: app/controllers/business_stats_controller.rb [hunk 25]
```diff
@@ -1,6 +1,15 @@
 class BusinessStatsController < ApplicationController
+  Route = Struct.new(:origin, :destination)
```
Comment (Recursively) [inline app/controllers/business_stats_controller.rb]: @TanaTanoi Thoughts on struct in controller lol
